### ISSUE TITLE
[Next Feature - ML] Automated Code Quality Improvement via Reinforcement Learning

### ISSUE THREAD

**Issue by thiagobutignon** [thiagobutignon/spec-kit-clean-archicteture#199]
## 🎯 Objetivo

Implementar agente de **Reinforcement Learning** que aprende a gerar e melhorar templates automaticamente, otimizando para scores RLHF mais altos.

## 📊 Contexto Atual

Templates são criados manualmente e melhorados através de feedback humano. Queremos um agente que:
- Aprenda a criar templates melhores
- Otimize automaticamente baseado em RLHF scores
- Proponha melhorias incrementais

### Reinforcement Learning Agent
- **Environment**: Sistema de geração de código
- **State**: Template atual + contexto (layer, target, features)
- **Actions**: Modificações no template (add pattern, remove import, etc)
- **Reward**: RLHF Score (-2 a +2)
- **Algorithm**: Proximal Policy Optimization (PPO) ou Deep Q-Network (DQN)

## 🏗️ Arquitetura Clean Architecture

```
features/ml-rl-optimizer/
├── domain/
│   ├── entities/
│   │   ├── rl-agent.ts            # Agente RL
│   │   ├── template-state.ts      # Estado atual
│   │   ├── improvement-action.ts  # Ação de melhoria
│   │   └── reward-signal.ts       # Sinal de recompensa
│   ├── usecases/
│   │   ├── train-rl-agent.ts
│   │   ├── improve-template.ts
│   │   └── evaluate-improvement.ts
│   └── repositories/
│       └── rl-model-repository.ts
├── data/
│   ├── usecases/
│   │   ├── ml-train-agent.ts
│   │   └── ml-improve-template.ts
│   └── protocols/
│       └── reward-calculator.ts
├── infra/
│   ├── ml/
│   │   ├── ppo-agent.ts           # PPO implementation
│   │   ├── replay-buffer.ts       # Experience replay
│   │   ├── policy-network.ts      # Neural policy
│   │   └── value-network.ts       # Value estimation
│   └── environment/
│       └── template-env.ts         # RL Environment
├── presentation/
│   └── cli/
│       ├── train-agent-command.ts
│       └── improve-command.ts
└── main/
    └── factories/
        └── rl-optimizer-factory.ts
```

## 📦 Tecnologias

1. **RL Framework**:
   - @tensorflow/tfjs-node (neural networks)
   - reinforcement-learning-js (RL algorithms)
   - stable-baselines3 (Python, para treinamento inicial)

2. **Training Infrastructure**:
   - Experience replay buffer
   - Multi-environment parallelization
   - Checkpoint system

### 1. Treinamento do Agente
```bash
regent rl train --episodes=1000 --layer=domain
# Output:
# 🤖 Training RL Agent...
#
# Episode 100/1000
#   Avg Reward: +0.8
#   Avg RLHF Score: +1.2
#   Improvements Found: 15
#
# Episode 500/1000
#   Avg Reward: +1.4
#   Avg RLHF Score: +1.7
#   Improvements Found: 42
#
# Training Complete!
#   Final Avg RLHF Score: +1.9
#   Best Template Saved: .regent/ml/rl/best-domain-template.regent
```

### 2. Melhoria Automática de Template
```bash
regent improve backend-domain-template.regent --auto
# Output:
# 🔧 Analyzing template...
# 🤖 RL Agent proposing improvements...
#
# Improvements Found:
#
# 1. Remove axios import (Score Impact: +0.5)
#    - Current: Violates domain layer
#    - Proposed: Use domain protocols instead
#
# 2. Add Value Object pattern (Score Impact: +0.3)
#    - Current: Primitive obsession
#    - Proposed: Wrap in Email, Password VOs
#
# 3. Improve error handling (Score Impact: +0.2)
#    - Current: Generic errors
#    - Proposed: Domain-specific error classes
#
# Estimated RLHF Score: +1.2 → +2.0
# Apply changes? [y/N]
```

### 3. Geração de Template do Zero
```bash
regent generate --feature="payment-processing" --use-rl
# Output:
# 🤖 RL Agent generating optimal template...
#
# Generated Template:
#   - Layer: domain, data, infra, presentation, main
#   - Patterns Used: Repository, Factory, Adapter, VO
#   - Predicted RLHF Score: +1.9
#   - Confidence: 92%
#
# Template saved: .regent/generated/payment-processing.regent
```

## 🔄 Training Loop

```typescript
// Episódio de treinamento
for (let episode = 0; episode < NUM_EPISODES; episode++) {
  // 1. Reset environment com template aleatório
  let state = env.reset()
  
  // 2. Loop até terminar episódio
  while (!done) {
    // 3. Agente escolhe ação (modificação no template)
    const action = agent.selectAction(state)
    
    // 4. Execute ação e obtenha recompensa
    const { nextState, reward, done } = env.step(action)
    
    // 5. Armazene experiência
    replayBuffer.add(state, action, reward, nextState, done)
    
    // 6. Treine agente com batch de experiências
    if (replayBuffer.size() > BATCH_SIZE) {
      agent.train(replayBuffer.sample())
    }
    
    state = nextState
  }
  
  // 7. Valide template final
  const finalScore = await rlhf.calculateScore(env.currentTemplate)
  
  console.log(`Episode ${episode}: RLHF Score = ${finalScore}`)
}
```

## 📈 Métricas de Sucesso

- **Learning Curve**: Score médio aumenta ao longo do treinamento
- **Convergence**: Estabilizar em score médio > +1.5
- **Generalization**: Funcionar bem em templates não vistos
- **Sample Efficiency**: Aprender com < 500 episódios
- **Impacto**: Templates gerados têm score +0.5 maior que baseline

### Fase 1: Environment (2 semanas)
- [ ] Criar RL environment
- [ ] Definir state/action/reward
- [ ] Sistema de validação automática
- [ ] Replay buffer

### Fase 2: Agent (3 semanas)
- [ ] Implementar policy network
- [ ] Implementar value network
- [ ] Algoritmo PPO/DQN
- [ ] Training loop

### Fase 3: Integration (2 semanas)
- [ ] CLI para treinamento
- [ ] Sistema de checkpoint
- [ ] Visualização de progresso
- [ ] Validação com templates reais

### Fase 4: Production (contínuo)
- [ ] Continuous learning
- [ ] Multi-task learning (várias layers)
- [ ] Transfer learning
- [ ] A/B testing de agentes

## 🔗 Dependências

- Issue #196: Predictive RLHF Score (para reward shaping)
- Issue #197: Pattern Detection (para action space)
- Issue #198: Recommendation System (para warm start)

### Reward Shaping
- **Dense Rewards**: Pequenas recompensas para cada melhoria
- **Sparse Rewards**: Grande recompensa apenas no final
- **Hybrid**: Combinar ambos

### Action Space
- Add/remove import
- Add/remove pattern
- Modify validation logic
- Change error handling
- Adjust file structure

### Exploration vs Exploitation
- Epsilon-greedy: 10% random actions
- Boltzmann exploration: Probabilidade baseada em Q-values
- Decaying epsilon ao longo do treinamento

### Safety
- Validar todas as modificações
- Rollback automático se score cair
- Human-in-the-loop para mudanças críticas
- Sandbox para execução segura
